node/src/Pos/Delegation/Logic/VAR.hs

```diff
@@ -556,23 +560,33 @@ dlgRollbackBlocks blunds = do
         let postedOp = SomeBatchOp $ map (GS.DelPostedThisEpoch . addressHash) issuers
         pure $ pskOp <> postedOp
 
--- | Normalizes the memory state after the rollback.
+-- | Normalizes the memory state after the rollback. Must be called
+-- only when 'StateLock' is taken.
 dlgNormalizeOnRollback ::
        forall ssc ctx m.
        ( MonadDelegation ctx m
        , DB.MonadBlockDB ssc m
        , DB.MonadGState m
        , MonadIO m
        , MonadMask m
-       , HasLens LrcContext ctx LrcContext
+       , HasLens' ctx LrcContext
        , Mockable CurrentTime m
+       , WithLogger m
+       , HasCoreConstants
+       , SscHelpersClass ssc
        )
     => m ()
 dlgNormalizeOnRollback = do
     tip <- DB.getTipHeader @ssc
     oldPool <- runDelegationStateAction $ do
-        dwEpochId .= (tip ^. epochIndexL)
         pool <- uses dwProxySKPool toList
         dwProxySKPool .= mempty
         pure pool
-    forM_ oldPool (processProxySKHeavy @ssc)
+    forM_ oldPool $ \psk -> do
+        res <- processProxySKHeavyInternal @ssc psk
+        -- should we throw error here also?
+        when (res /= PHAdded) $ logError $
+            sformat ("dlgNormalizeOnRollback: got psk verdict "%shown%" for psk "%build)
+                    res
+                    psk
+    runDelegationStateAction $ dwTip .= headerHash tip
```

Comment:
All this logic looks incorrect. When we call `dlgNormalizeOnRollback`, we have already updated tip in DB, but haven't updated it in mempool. So I expect `processProxySKHeavyInternal` to always fail due to tip mismatch.
You should update mempool and set tip inside single atomic STM action, otherwise it can break invariants.